lib/ecto/repo/preloader.ex

```diff
@@ -272,6 +272,19 @@ defmodule Ecto.Repo.Preloader do
     We expected a tuple but we got: #{inspect(entry)}
     """
 
+  defp related_key_to_field(%Ecto.Association.ManyToMany{} = assoc, query, {pos, key}) do
```

Comment:
We shouldn't directly access any association here because this code should be association agnostic. Instead, you should most likely change `preload_info` to return an expanded contract that looks like this:

    {:assoc, refl, {-1, join_owner_key, type}}

If the three element tuple is returned, you wrap it with the given type.

Reply:
Agreed this is a much cleaner approach. I just implemented this and the tests are passing (aside from those myxql ones).